### DISCUSSION TITLE
ComfyUI (image generation) and AllTalk (TTS)

### DISCUSSION THREAD

**Discussion by gibru** [open-webui/open-webui#1221] [Ideas]
Playing with image generation inside of OpenWebUI using A1111 as the engine (which works well btw), I was wondering whether it'd be possible (feasible?) to add [ComfyUI](https://github.com/comfyanonymous/ComfyUI) as an engine. It works really well powering the IMO excellent [krita-ai-diffusion](https://github.com/Acly/krita-ai-diffusion) plugin and, at least on my machine, runs way faster than A1111.

Another feature that might be worth looking into is integration of [AllTalk TTS](https://github.com/erew123/alltalk_tts) for things like TTS. In my testing it delivers pretty good quality (and in some of my tests I'd say even excellent).

Of course, just some suggestions.

**Comment by Michelklingler** [comment]
Adding ComfyUI would be Huge for US as well

**Comment by tjbck** [comment]
Basic ComfyUI integration coming soon!

**Comment by tjbck** [comment]
ComfyUI support has been added to our dev branch, please let me know if you encounter any issues!

**Reply by Michelklingler** [reply]
Amazing!
Looking forward to try this next week!

**Reply by ZaneA** [reply]
Thank you, it's working very well here! Today I tried Open WebUI for the first time, looked into ComfyUI support, and it's now here within the space of a few hours!

I'd love the ability to set a custom JSON/workflow which should also solve any requests for deeper configuration (e.g. sampler), it would require the user to provide node IDs for the payload (or perhaps a node ID + field name for each payload option so that even non-standard prompt nodes can work for example) but I think this could greatly increase capability and quality of outputs.

Something to consider anyway, even the basic support is great to have now. Thanks again for your work on this (and Open WebUI)! :)

**Reply by gibru** [reply]
Thanks a lot! Connecting to ComfyUI works without any problems. However, I'm running into an issue when using SDXL models with a standard resolution of 1024x1024: running out of memory. For reference: GPU is a an RTX 3060 with 12GB of VRAM. Using SD1.5 models at a 512x512 works without any problems.

By contrast, when switching to the A1111 engine, SDXL generations at 1024x1024 work flawlessly.

**Reply by Michelklingler** [reply]
I agree with Zena. Being able to setup a custom JSON would incredible.

**Reply by Michelklingler** [reply]
![image](https://github.com/open-webui/open-webui/assets/63208430/21741c47-7330-4228-b0c5-90885f86fb4a)

But I guess it would also be possible to copy paste the JSON here referring the prompt and model variable as a temporary workaround

**Reply by nattyraz** [reply]
Hi ! nice can you add in the documentation, juste the way for set up this  ? cause i am a noob & i can not find the way to use it .... i can set up easy SD 1111 but not comfyUi...

**Comment by Michelklingler** [comment]
Could you monitor your GPU memory load?
SDXL is already about 8GB if you load a LLM at the same time it is tight.
And make sense that you would run out of vRam

**Reply by gibru** [reply]
That's a perfectly reasonable assumption, yes. However, as I said, when using AUTOMATIC1111 as the engine I have no issues with SDXL at 1024x1024. Maybe something in the ComfyUI configs I should look for?

**Comment by Michelklingler** [comment]
I know that Comfy tend to use a little more vram as automatic. Especially
if in the workflow you have additional Lora, upscaler or any other extra
model that might add up to your load. Did you try with a simple default
workflow or is it a custom one?

**Comment by gibru** [comment]
> I know that Comfy tend to use a little more vram as automatic.

Thanks for the hint! I fixed the issue by starting comfy in lowvram mode:

`python main.py --lowvram`

**Comment by shuasimodo** [comment]
I second AllTalk TTS. I think the biggest advantages are: 

1. it can alternate between ram and vram
2. you can create your own voices by simply placing a .wav file into the directory. 
3. 99% local. (There is a fine-tuning issue that they mention in docs how to disable.)
4. works pretty seamlessly
5. built in fine-tuning for custom voices

There are some memory issues though. I found it often running out of memory but it could have been more of a text-generation-webui memory management issue than anything else. 
CONTEXT: 
in comparison between ollama:openwebui running whisper, and oobabooga running whisper, I found oobabooga uses vram after you speak to it, here does not.
- If that same ram management can be worked into the current setup then you'll have the perfect trifecta

for anyone curious, the quickest way to get it up and running fully is oobabooga AFAIK. I think there's also a model that gets downloaded after the installation, when you activate the extension and apply the changes. 
https://github.com/erew123/alltalk_tts/?tab=readme-ov-file#-quick-setup-text-generation-webui--standalone-installation

**Reply by gibru** [reply]
> for anyone curious, the quickest way to get it up and running fully is oobabooga AFAIK.

standalone installation is as quick (I've installed it both ways to experiment a little). And  for those who don't have oobabooga already installed it'd be even faster ;) That said and just out of curiosity since you occasionally run out of memory, have you ever tried the standalone installation to check whether it's AllTalk or, as you suspect, an oobabooga issue? 

Also, as I keep looking for and experimenting with different tools, I recently discovered another TTS solution that looks promising: [VoiceCraft](https://github.com/jasonppy/VoiceCraft) (model weights are not available yet, though). The [demo](https://jasonppy.github.io/VoiceCraft_web/) sounds pretty good (with your typical comparisons including XTTS v2).

**Reply by shuasimodo** [reply]
NOPE. Never ventured down the tinkering path until now. I'm going to play with the standalone today and set up a couple other local TTS if i get time. 

will report back

reporting back: standalone isn't any better.

**Comment by GenieDev101** [comment]
It is my understanding that both AllTalk and VoiceCraft would likely affect the License of Open WebUI, and I would suggest considering the different licenses of any implementations of other projects and making sure the required license changes are desirable before they are implemented into Open WebUI

**Reply by justinh-rahb** [reply]
Correct, AllTalk is AGPL-licensed, which means we'd need to relicense to a compatible GPL license.

**Comment by slash-proc** [comment]
Coqui/AllTalk can be used if they run as their own instance separately. There shouldn't be any license issues since they have api endpoints we can use. I imagine the backend would need to handle getting the TTS data from AllTalk and then pass it to the frontend. 

I'm not a frontend guy so I don't know how that would fit into the current way of handling TTS. From what I can tell it's all done client side, right? Either built-in browser support (obviously local) or using openai (but all in the frontend?)

**Reply by GenieDev101** [reply]
I do not believe your idea is compatible with the AGPL license

**Reply by slash-proc** [reply]
How is allowing open-webui to communicate with a 3rd party api non compliant with the AGPL? Since it's not being packaged with open-webui that shouldn't matter.

[Here's the endpoint for AllTalk](https://github.com/erew123/alltalk_tts/blob/66dadbca251115609e006051177a8704520de115/tts_server.py#L601)

**Reply by GenieDev101** [reply]
You are right. I am sorry. I did some further research, and you are correct

**Comment by edurenye** [comment]
For TTS, STT and wake words we could integrate something like the [Wyoming Protocol](https://github.com/rhasspy/wyoming) allowing us to connect it to many TTS like Whisper or Vosk, STT like Piper and Wake word detection software like openWakeWord, porcupine1 or snowboy.
